add code of conduct and contributing guide (#83)
insightsengineering/r.pkg.template: CONTRIBUTING.md

# Contributor guide

First of all, thank you for taking the time to contribute!✌️ 🎉

We love your input regardless of its size, content or scope - whether it is an issue, feature request, bug-fix or documentation enhancement.

Thanks 🙏 You rock 🤘

## Table Of Contents

[Getting started](#getting-started-👶)

[Code of Conduct](#code-of-conduct)

[License](#license)

[How to report a bug?](#how-to-report-a-bug-🐛)

[How to request the change?](#how-to-request-the-change-💡)

[How to submit a change?](#how-to-submit-a-change-🚩)

[Style guide](#style-guide-👗)

[Recognision model](#recognition-model-🧐)

[Any questions?](#any-questions-❓)

## Getting started 👶

> For an introduction, please read [`README`](README.md) file. You can also find vignettes [here](vignettes/). Most likely you would find a "Getting started" article. Please also see a package manual in form of `pkgdown` website.

Reviewer: ```suggestion
For an introduction, please read [`README`](README.md) file. You can also find vignettes [here](vignettes/). Usually you will find a "Getting started" article. Please also see the package manual in the form of a `pkgdown` website.
```

## Code of Conduct

This project is governed by [Code of Conduct](CODE_OF_CONDUCT.md). By participating, you are expected to follow the rules outlined there.

## License

All of your code changes would be under the same [license](LICENSE) that covers this project.

## How to report a bug 🐛 ?

We use GitHub to track issues, feature requests as well as bugs. Before opening a new issue please double-check if it's already reported (but don't worry if  to be a duplicate - we will manage). If already there - up-vote ⬆️. We will have a look at it faster!

> We kindly ask you to write a good issue with a minimal and reproducible example that showoff the problem. It's would be also great to know your local environment (R session info) as well. You will find our guides in the template while opening a bug ticket. This would speed up the fixing process a lot! 📈.

Reviewer: ```suggestion
We kindly ask you to write a good issue with a minimal and reproducible example that demonstrates the problem. It would also be great to know your local environment (R session info) as well. You will find our template for issues when opening a bug ticket. This will speed up the fixing process a lot! 📈.
```

## How to request the change 💡 ?

> In case you find a missing feature or you have an idea how to enhance existing functionalities - please let us know by creating a new issue. Before opening a new issue please double-check if it's already reported (but don't worry if occurs to be a duplicate - we will manage). If already there - up-vote ⬆️. We will have a look at it faster!

Reviewer: ```suggestion
If you notice a missing feature or you have an idea how to enhance existing functionalities - please let us know by creating a new issue. Before opening a new issue please double-check if it's already reported (but don't worry if occurs to be a duplicate - we will manage). If already there - up-vote ⬆️. We will have a look at it faster!
```

> Please elaborate on "why?" - what's the context, what's the benefit and for whom it is. You will find our guides in the template while opening a feature request ticket. This would help us prioritize and submit meaningful changes.

Reviewer: ```suggestion
Please elaborate on "why?" - what's the context, what's the benefit and for whom. You will find our template when opening a feature request ticket. This will help us prioritize and submit meaningful changes.
```

## How to submit a change 🚩 ?

### GitHub Flow

We are following [GitHub Flow](https://docs.github.com/en/get-started/quickstart/github-flow) to collaborate in this repository. In order to submit the changes please do the following:

1. Open a branch

>     Please see branching convention below. You can open a branch on remote as well as a fork.

Reviewer: Are we opening the repo to write access or are we making external collaborators open a fork (I think the latter is more usual?) @dinakar29

Author: I will rephrase it into the following
"If you don't have write access please fork it first."

1. Make changes.

>      Make sure your code is passing all the checks, it is well documented and tested so as not to decrease existing test coverage 💪.

Reviewer: ```suggestion
     Make sure your code passes all the checks and is well documented and tested so as not to decrease existing test coverage 💪.
```

1. Create a PR(s)

    Please link an issue and make a good change description. Include development assumptions when necessary. Give all the details for a reviewer to efficiently check your changes.

1. Address review comments

    Please apply changes where necessary. If you disagree with a reviewer - please explain why.

> 1. Merge your PRs and delete a branch

Reviewer: Do you we want to give control to contributors to do the merge - feels like it's something we should do?

Author: once approved by any of the codeowners it should be ok

### Branching convention

In case you are working on a task inside one specific repository, please name your branch `<issue_id>_<short_description>` all lowercase. Multiple words of the description should be divided by an underscore (`_`). E.g. `15_fix_spelling_error` in case you try to solve a spelling mistake mentioned in the issue number `15`.

In case you are working on a task from one repository that affects multiple repositories, please always
name your branches: `<issue_id>_<issue_repo>_<short description>` all lowercase. Multiple words of the description should be divided by an underscore (`_`). E.g. `15_abc_fix_spelling_error` in case you try to solve a spelling mistake inside `xyz` which closes issue `15` inside `abc`.

### Monorepo and staged.dependencies

> Sometimes in order to introduce a change you also need a change from an upstream dependent package. We are using [staged.dependencies](https://github.com/openpharma/staged.dependencies) functionality to simulate a monorepo behavior. The configuration is already in the [staged_dependencies.yaml](staged_dependencies.yaml) file. You just need to name the feature branches identically to take advantage of that. Please read the package manual for more details.

Reviewer: So this seems a little confusing as if you end up using staged.deps you explicitly do not follow the above branching convention

### Recommended development environment & tools

#### R & package versions

> We continuously test our packages against the newest R version as well as a given package dependencies. We recommend to set-up your working environment in the same way. You can find all the details in any of the actions.

Reviewer: Link to action? Or say "see below"

If you find out any bugs on the older version of dependencies - please create appropriate bug ticket.

#### pre-commit

We highly recommend the [`pre-commit`](https://pre-commit.com/) tool combined with [`R hooks for pre-commit`](https://github.com/lorenzwalthert/precommit) to execute some of the checks prior committing and pushing. The configuration is already there in a repo ([here](.pre-commit-config.yaml)). Please, follow the installation guide on the official [`pre-commit` page](https://github.com/lorenzwalthert/precommit) and the [`GitHub` `readme` page](https://github.com/lorenzwalthert/precommit#installation) for the R hooks.

## Style guide 👗

This repository follows standard [`tidyverse` style guide](https://style.tidyverse.org/) and it's being checked against it by [`lintr`](https://github.com/r-lib/lintr). There are some slight modifications to its default settings available in the [`.lintr`](.lintr) file.

> Although it allows for some flexibility - we recommend to stick to the style already present in the existing code. At the end of the day we want to have consistent codebase.

Reviewer: ```suggestion
Although it allows for some flexibility - we recommend sticking to the style of the existing code. At the end of the day we want to have consistent codebase.
```

> Please note that there is a style and also `lintr` check in place that would validate your code.

Reviewer: ```suggestion
Please note that there is a style and also `lintr` check in place that will validate your code.
```

# Contributor guide

> First of all, thank you for taking the time to contribute!✌️ 🎉

Reviewer 2: ```suggestion
First of all, thank you for taking the time to contribute! ✌️ 🎉
```

We love your input regardless of its size, content or scope - whether it is an issue, feature request, bug-fix or documentation enhancement.

Thanks 🙏 You rock 🤘

> ## Table Of Contents

Reviewer 2: ```suggestion
## Table of contents
```

### Recommended development environment & tools

#### R & package versions

We continuously test our packages against the newest R version as well as a given package dependencies. We recommend to set-up your working environment in the same way. You can find all the details in a given GitHub Action execution log - there is a step that prints out session info.

If you find out any bugs on the older version of dependencies - please create appropriate bug ticket.

#### `pre-commit`

We highly recommend the [`pre-commit`](https://pre-commit.com/) tool combined with [`R hooks for pre-commit`](https://github.com/lorenzwalthert/precommit) to execute some of the checks prior committing and pushing. The configuration is already there in a repository ([here](.pre-commit-config.yaml)). Please, follow the installation guide on the official [`pre-commit` page](https://github.com/lorenzwalthert/precommit) and the [`GitHub` `readme` page](https://github.com/lorenzwalthert/precommit#installation) for the R hooks.

## Style guide

> This repository follows standard [`tidyverse` style guide](https://style.tidyverse.org/) and it's being checked against it by [`lintr`](https://github.com/r-lib/lintr). There are some slight modifications to its default settings available in the [`.lintr`](.lintr) file.

Reviewer 3: `spelling` is also used here

Author: ... and many others. But those are not focusing on styling the code (see header). I didn't want to list all of our checks as that list is too dynamic

Although it allows for some flexibility - we recommend sticking to the style of the existing code.

Please note that there is a style and also `lintr` check in place that will validate your code.

## Recognition model

Any contribution is highly welcomed and appreciated. While all the contribution data is already there in GitHub repository insights feature, we introduced some objective rules to recognize a _significant_ contribution so as to became a package author:

- Minimum 5% of lines of code authored (determined by `git blame` query) OR
- Being at the top 5 contributors in terms of number of commits OR lines added OR lines removed.

> The package maintainer also reserve the rights to manually adjust the criteria to recognize contributions.

Reviewer 2: ```suggestion
The package maintainer also reserves the rights to manually adjust the criteria to recognize contributions.
```

## Any questions

If you have any further questions regarding contribution - please reach out to the repository maintainer!

## Note

> This is a template applied across multiple repositories within our organization. You might find additional, repository specific rules in the [`CONTRIBUTING_ADDITIONAL.md`](CONTRIBUTING_ADDITIONAL.md) file. If it's missing then it means that there is no additional considerations to the above.

Author: Does it make sense for you guys? Do you think we need it?
To me, there are some repo-specific rules e.g. in tern (specific naming convention, templating, etc.). I just want to open a door to specifying this while keeping the base identical across the repos. Will you use it or it's too much? @shajoezhu @gogonzo

Reviewer 2: I wonder if we should also keep a copy of `CONTRIBUTING_ADDITIONAL.md` in this repo as well. Just to keep consistency. And suggest for removal if not needed.

Reviewer 3: Can't we just edit CONTRIBUTING.md in the future in the specific repository instead of making new files?

Author: > I wonder if we should also keep a copy of CONTRIBUTING_ADDITIONAL.md in this repo as well. Just to keep consistency. And suggest for removal if not needed.

I would rather have opt-in strategy than opt-out. There are too many repos to manage - we would forget about something.

> Can't we just edit CONTRIBUTING.md in the future in the specific repository instead of making new files?

Yeah - that's definitively an option. I was thinking about that and this would block future potential mass-updates of that file (if there would be a need). Therefore I suggested an additional, optional, repo-specific file. I know that this is something unusual but at least we would have a shared, master version that would be easy to update. Does that convince you or it's too much?

Author: After I second thought I am going to remove that section because of YAGNI. Let's start with more simpler version and then we would add it when we need it.

## Style guide

This repository follows standard [`tidyverse` style guide](https://style.tidyverse.org/) and it's being checked against it by [`lintr`](https://github.com/r-lib/lintr). There are some slight modifications to its default settings available in the [`.lintr`](.lintr) file.

Although it allows for some flexibility - we recommend sticking to the style of the existing code.

Please note that there is a style and also `lintr` check in place that will validate your code.

## Recognition model

Any contribution is highly welcomed and appreciated. While all the contribution data is already there in GitHub repository insights feature, we introduced some objective rules to recognize a _significant_ contribution so as to became a package author:

- Minimum 5% of lines of code authored (determined by `git blame` query) OR
- Being at the top 5 contributors in terms of number of commits OR lines added OR lines removed.

The package maintainer also reserve the rights to manually adjust the criteria to recognize contributions.

## Any questions

If you have any further questions regarding contribution - please reach out to the repository maintainer!

## Note

> This is a template applied across multiple repositories within our organization. You might find additional, repository specific rules in the [`CONTRIBUTING_ADDITIONAL.md`](CONTRIBUTING_ADDITIONAL.md) file. If it's missing then it means that there is no additional considerations to the above.

Author: Does it make sense for you guys? Do you think we need it?
To me, there are some repo-specific rules e.g. in `tern` (specific naming convention, templating, etc.). I just want to open a door to specifying this while keeping the base identical across the repos. Will you use it or it's too much? @shajoezhu @gogonzo
